simpleclient_common/src/main/java/io/prometheus/client/exporter/common/MetricMapper.java

```diff
@@ -0,0 +1,252 @@
+package io.prometheus.client.exporter.common;
+
+import org.yaml.snakeyaml.Yaml;
+
+import java.io.Reader;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static java.lang.String.format;
+
+/**
+ * Map origin names with a target name, label names and values according a specified configuration.
+ */
+public class MetricMapper {
+
+    private static final Pattern snakeCasePattern = Pattern.compile("([a-z0-9])([A-Z])");
+    private final boolean lowerCaseOutputNames;
+    private final boolean lowerCaseOutputLabelNames;
+    private final Pattern unsafeChars = Pattern.compile("[^a-zA-Z0-9:_]");
+    private final Pattern multipleUnderscores = Pattern.compile("__+");
+    Map<String, MetricMapping> mappingCache;
+    private ArrayList<Rule> rules = new ArrayList<Rule>();
+
+    /**
+     * @param rules                    a list of mapping rules to apply.
+     * @param lowerCaseOutputNames     lowercase metric names.
+     * @param lowerCaseOutpuLabelNames lowercase metric labels names.
+     */
+    public MetricMapper(ArrayList<Rule> rules, boolean lowerCaseOutputNames, boolean lowerCaseOutpuLabelNames) {
+        this.rules = rules;
+        this.mappingCache = new HashMap<String, MetricMapping>();
+        this.lowerCaseOutputNames = lowerCaseOutputNames;
+        this.lowerCaseOutputLabelNames = lowerCaseOutpuLabelNames;
+    }
+
+    /**
+     * Replace invalid chars to underscore and replace multiple underscores with one underscore.
+     *
+     * @param s a metric name or label name.
+     * @return a sanitized name.
+     */
+    private String safeName(String s) {
+        // Change invalid chars to underscore, and merge underscores.
+        return multipleUnderscores.matcher(unsafeChars.matcher(s).replaceAll("_")).replaceAll("_");
+    }
+
+    /**
+     * Map a metric to target mapping.
+     *
+     * @param metricName
+     * @return a mapping for the specified metric name.
+     */
+    public MetricMapping map(String metricName) {
+        if (!mappingCache.containsKey(metricName)) {
+            mappingCache.put(metricName, process(metricName));
+        }
+        return mappingCache.get(metricName);
+    }
+
+    /**
+     * Process rules for the specified metric name and map it with label names, values and
+     * help according to the first matched rule.
+     *
+     * @param metricName
+     * @return a MetricMapping associated with a metricName.
+     */
+    public MetricMapping process(String metricName) {
+        String targetMetricName;
+        final String targetHelp;
+        final String snakeCaseMetricName = snakeCasePattern.matcher(metricName).replaceAll("$1_$2").toLowerCase();
+
+        for (Rule rule : rules) {
+            Matcher matcher = null;
```

Comment:
Maybe move the inside of the loop in a sub-method ? (This will allow you to write a unit test to test only one iteration)